### ISSUE TITLE
Rule Change: no-param-reassign to catch reassignment via Object.assign()

### ISSUE THREAD

**Issue by adamni21** [eslint/eslint#16915]
### What rule do you want to change?

no-param-reassign

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
const reassign = (obj) => {
  Object.assign(obj, { message: "reassigned" });
  return "something else"
};
```

### What does the rule currently do for this code?

Nothing.

### What will the rule do after it's changed?

Give a no-param-reassign warning.

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by Rec0iL99** [comment]
Hi @adamni21, thanks for opening the issue. This is a duplicate of #7570 and the proposal then was rejected by the team because there are a lot of ways an object can be mutated (using `Object.assign`, `Reflect.set`, etc), so you cannot really prioritize what to restrict and what not to.

An original idea by @fasttime:

> Just to summarize:

> ```js
> obj["foo"] = 1;                                       // error
> obj.foo = 1;                                          // error
> Object.assign(obj, { foo: 1 });                       // error
> Reflect.set(obj, "foo", 1);                           // ok
> Object.defineProperty(obj, "foo", { value: 1 });      // ok
> Object.defineProperties(obj, { foo: { value: 1 } });  // ok
> ```

The original issue is old and a lot has changed then, I'm moving this for more feedback.

**Comment by mdjermanovic** [comment]
> there are a lot of ways an object can be mutated (using `Object.assign`, `Reflect.set`, etc), so you cannot really prioritize what to restrict and what not to.

I agree with this, we cannot entirely prevent modifying objects so I'm not much in favor of adding checks for some particular ways, though I wouldn't be opposed to this change if other team members support it. However, if accepted, I think it would require adding a new option.

**Comment by adamni21** [comment]
I'm not quite sure, whether you waiting for a comment from my side, so just in case.

I don't feel very strongly this.

I just caught myself working around the linter, doing the not recommended thing in different way, to not get a warning.
So, I just wanted to point that out.
And unfortunately, didn't find the already existing (github)-issue, sorry for that.

But I find the point by @fasttime very reasonable.

Btw, thanks for your incredible work.

**Comment by nzakas** [comment]
I'm in agreement with @mdjermanovic - I don't think it makes sense to make a change here, so closing.
